src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DetectAmbiguousRoutes.cs

```diff
@@ -0,0 +1,116 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Analyzers.Infrastructure;
+using Microsoft.AspNetCore.Analyzers.Infrastructure.RoutePattern;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers.RouteHandlers;
+
+public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
+{
+    private static void DetectAmbiguousRoutes(in OperationBlockAnalysisContext context, List<MapOperation> blockRouteUsage)
+    {
+        if (blockRouteUsage.Count == 0)
+        {
+            return;
+        }
+
+        var groupedByParent = blockRouteUsage
+            .Where(u => !u.RouteUsageModel.UsageContext.HttpMethods.IsDefault)
+            .GroupBy(u => new MapOperationGroupKey(u.Operation, u.RouteUsageModel.RoutePattern, u.RouteUsageModel.UsageContext.HttpMethods));
+
+        foreach (var ambigiousGroup in groupedByParent.Where(g => g.Count() >= 2))
```

Comment:
Ah, so you are grouping invocations by common parenting block operation and analyzing each group. Will this handle invocations within conditional code that is not parented by a block operation? For example:

```csharp
void M(bool flag)
{
   InvocationToAnalyze();
   var result = flag ? InvocationToAnalyze() : InvocationToNotAnalyze();
}

```

Both `InvocationToAnalyze` calls above share the same parenting `IBlockOperation`, but it is not guaranteed that both calls will always be made on all code paths.